File: Sudoku_Solver.py
# unsolved Sudoku Board
board = [
    [7,8,0,4,0,0,1,2,0],
    [6,0,0,0,7,5,0,0,9],
    [0,0,0,6,0,1,0,7,8],
    [0,0,7,0,4,0,2,6,0],
    [0,0,1,0,5,0,9,3,0],
    [9,0,4,0,6,0,0,0,5],
    [0,7,0,3,0,0,0,1,2],
    [1,2,0,0,0,7,4,0,0],
    [0,4,9,2,0,6,0,0,7]
]

def solve(bo):
    """
    Solve the given sudoku using recursion and backtracking
    :param bo: Board
    :return:
    """
    find = find_empty(bo)
    # Base case
    if not find:
        return True
    else:
        row,col = find

    # Looping through the values from 1 to 9
    for i in range(1,10):

        # Check if the solution is valid or not
        if valid(bo,i,(row,col)):

            # If valid add it into the board
            bo[row][col] = i

            # Recursively call solve
            if solve(bo):
                return True

            # If solve = False Backtrack and reset the last element
            bo[row][col] = 0

    return False

def valid(bo,num,pos):
    """
    Given the board check if it is valid or not .
    We will check here the row , the column and the inner square .
    :param bo: Board
    :param num: Inserted value
    :param pos: (i,j) => The Location of empty cells
    """
    # Check row
    for i in range(len(bo[0])):
        if bo[pos[0]][i] == num and pos[1] != i:
            return False

    # Check columns
    for i in range(len(bo)):
        if bo[i][pos[1]] == num and pos[0] != i:
            return False

    # Figure out which box we are in
    """
    7 8 0  | 4 0 0  | 1 2 0
    6 0 0  | 0 7 5  | 0 0 9
    0 0 0  | 6 0 1  | 0 7 8
    - - - - - - - - - - - - - 
    0 0 7  | 0 4 0  | 2 6 0
    0 0 1  | 0 5 0  | 9 3 0
    9 0 4  | 0 6 0  | 0 0 5
    - - - - - - - - - - - - - 
    0 7 0  | 3 0 0  | 0 1 2
    1 2 0  | 0 0 7  | 4 0 0
    0 4 9  | 2 0 6  | 0 0 7
    """
    box_x = pos[1] // 3  # Eg. 0,1,2
    box_y = pos[0] // 3  # Eg. 0,1,2

    # Looping through every box
    for i in range(box_y * 3, box_y*3 + 3):
        for j in range(box_x * 3, box_x*3 + 3):
            if bo[i][j] == num and (i,j) != pos:
                return False

    # If we make it through every box then we will return True
    return True

def print_board(bo):
    """
    Print the given/ sudoku board
    :param bo: Board
    """
    for i in range(len(bo)):  # Rows => 9
        if i % 3 == 0 and i != 0:
            print("- - - - - - - - - - - - - ")

        for j in range(len(bo[0])):  # Columns => 9
            if j % 3 == 0 and j != 0:
                print(" | ", end="")

            if j == 8:
                print(bo[i][j])
            else:
                print(str(bo[i][j]) + " ", end="")


def find_empty(bo):
    """
    Count the no of enpty cells in a given sudoku board
    :param bo: Board
    """
    for i in range(len(bo)): # Rows
        for j in range(len(bo[0])): # Columns
            if bo[i][j] == 0:
                return (i, j)  # row, col

    return None

print_board(board)
solve(board)
print("\n")
print("Solved Sudoku:")
print_board(board)

File: Sudoku_Solver.cpp
#include<iostream>
#include <bits/stdc++.h>
using namespace std;

#define U 0
#define N 9

// Finding 0 in every row and col
bool FindUnassignedLoc(int grid[N][N], int& row,int& col){
    for(row =0;row<N;row++){
        for(col=0;col<N;col++){
            if(grid[row][col]== U )
                return true;
        }
    }
    return false;
}

// Check if the assigned no. is placed in its correct row
bool usedInRow(int grid[N][N],int row,int num){
    
        for(int col = 0;col<N;col++){
            if(grid[row][col]==num)
                return true;
        }
    
    return false;
}

// Check if the assigned no. is placed in its correct column
bool usedInCol(int grid[N][N],int col,int num){
    
        for(int row = 0;row<N;row++){
            if(grid[row][col]==num){
                return true;
            }
        }
    
    return false;
}

// Check if the assigned no. is in the correct grid
bool usedInBox(int grid[N][N],int box_row , int box_col,int num){
    for(int row=0;row<3;row++){
        for(int col=0;col<3;col++){
            if(grid[row+box_row][col+box_col]==num)
                return true;
        }
    }
    return false;
}

// Check if the no. is assigned to its accurate spot
bool isSafe(int grid[N][N],int row,int col,int num){
    return !usedInRow(grid,row,num) && !usedInCol(grid,col,num) && !usedInBox(grid,row-row%3,col-col%3,num) && grid[row][col]==U;
}

// Main Soduku solver function
bool Sudoku_solver(int grid[N][N]){
    int row,col;

    // Base Condition
    if(!FindUnassignedLoc(grid,row,col))
        return true;

    for(int num=1;num<=9;num++){

        if(isSafe(grid,row,col,num)){

            //Assign value to the empty box if the above condition return true
            grid[row][col]=num;

            if(Sudoku_solver(grid))
                return true;

            //If above case fail to return true
            grid[row][col] = U;
        }
    }
    return false;
}

// Print the resultant gird
void print_grid(int grid[N][N]){
    for(int row=0;row<N;row++){
        for(int col=0;col<N;col++){
            cout<<grid[row][col]<<" ";
        }
        cout<<endl;
    }
}


int main(){

    int grid[N][N] = {{7,8,0,4,0,0,1,2,0},
                      {6,0,0,0,7,5,0,0,9},
                      {0,0,0,6,0,1,0,7,8},
                      {0,0,7,0,4,0,2,6,0},
                      {0,0,1,0,5,0,9,3,0},
                      {9,0,4,0,6,0,0,0,5},
                      {0,7,0,3,0,0,0,1,2},
                      {1,2,0,0,0,7,4,0,0},
                      {0,4,9,2,0,6,0,0,7}};

    cout<<"-------SUDOKU BOARD UNSOLVED--------"<<endl;
    print_grid(grid);

    if(Sudoku_solver(grid)==true){
        cout<<"-------SUDOKU BOARD SOLVED--------"<<endl;
        print_grid(grid);
    }
    else{
        cout<<"No Solution exist";
    }

    return 0;

}
